### ISSUE TITLE
[BUG] Learning engine not support  categorical_attrs_desc

### ISSUE THREAD

**Issue by acezen** [alibaba/GraphScope#503]
**Describe the bug**
   when use training config with categorical_attrs_desc, training would failed. seems learning engine not support  categorical_attrs.

**To Reproduce**
```python
def train(config, graph):
  def model_fn():
    return  BipartiteGraphSage(graph,
                               config['batch_size'],
                               config['hidden_dim'],
                               config['output_dim'],
                               config['hops_num'],
                               config['u_neighs_num'],
                               config['i_neighs_num'],
                               u_features_num=config['u_features_num'],
                               u_categorical_attrs_desc=config['u_categorical_attrs_desc'],
                               i_features_num=config['i_features_num'],
                               i_categorical_attrs_desc=config['i_categorical_attrs_desc'],
                               e_features_num=config['e_features_num'],
                               e_categorical_attrs_desc=config['e_categorical_attrs_desc'],
                               neg_num=config['neg_num'],
                               use_input_bn=config['use_input_bn'],
                               act=config['act'],
                               agg_type=config['agg_type'],
                               need_dense=config['need_dense'],
                               in_drop_rate=config['drop_out'],
                               ps_hosts=config['ps_hosts'])
  trainer = LocalTFTrainer(model_fn,
                              epoch=config['epoch'],
                              optimizer=gl.get_tf_optimizer(
                                  config['learning_algo'],
                                  config['learning_rate'],
                                  config['weight_decay']))

  trainer.train_and_evaluate()

  epoch_num = config['epoch']
  u_embs = trainer.get_node_embedding("u")
  np.save(f'../embedding/2hopsu_emb{epoch_num}_epoch', u_embs)

  i_embs = trainer.get_node_embedding("i")
  np.save(f'../embedding/2hopsi_emb{epoch_num}_epoch', i_embs)

def main():
    config = {'batch_size': 2048,
            'hidden_dim': 64,
            'output_dim': 64,
            'u_features_num': 3,
            'u_categorical_attrs_desc': {"0":["u_id",100000,64],},
            'i_features_num': 3,
            'i_categorical_attrs_desc': {"0":["i_id",100000,64]},
            "e_features_num":4,
            "e_categorical_attrs_desc":{},
            'hops_num': 2,
            'u_neighs_num': [5, 2],
            'i_neighs_num': [5, 2],
            'neg_num': 10,
            'learning_algo': 'adam',
            'learning_rate': 0.001,
            'weight_decay': 0.000,
            'epoch': 1,
            'use_input_bn': True,
            'act': tf.nn.leaky_relu,
            'agg_type': 'gcn',
            'need_dense': True,
            'drop_out': 0.0,
            'ps_hosts': None}

    vertices = {
        "u": (Loader("/home/jovyan/datasets/amazon-book/scope_user.csv", delimiter=",",), ["idfeature:string", "degree:int64","sqrtdegree:double"], ),
        "i": (Loader("/home/jovyan/datasets/amazon-book/scope_item.csv", delimiter=","), ["idfeature:string", "degree:int64","sqrtdegree:double"], ),
        }
    edges = {
        "u-i": (
            Loader("/home/jovyan/datasets/amazon-book/scope_train.csv", delimiter=","),
            [ "src_degree:int64", "dst_degree:int64", "src_sqrtdegree:double", "dst_sqrtdegree:double", "label"],
            ("src_id:int64", "u"),
            ("dst_id:int64", "i"),
            # "label:int64"
        ),
        "u-i_reverse": (
            Loader("/home/jovyan/datasets/amazon-book/scope_train.csv", delimiter=","),
            [ "src_degree:int64", "dst_degree:int64", "src_sqrtdegree:double", "dst_sqrtdegree:double", "label"],
            ("dst_id:int64", "i"),
            ("src_id:int64", "u"),
            # "label:int64"
        ),
        "test-u-i": (
            Loader("/home/jovyan/datasets/amazon-book/scope_train.csv", delimiter=","),
            [ "src_degree:int64", "dst_degree:int64", "src_sqrtdegree:double", "dst_sqrtdegree:double", "label"],
            ("src_id:int64", "u"),
            ("dst_id:int64", "i"),
            # "label:int64"
        ),
        "test-u-i_reverse": (
            Loader("/home/jovyan/datasets/amazon-book/scope_train.csv", delimiter=","),
            [ "src_degree:int64", "dst_degree:int64", ("src_sqrtdegree:double") "dst_sqrtdegree:double", "label"],
            ("dst_id:int64", "i"),
            ("src_id:int64", "u"),
            # "label:int64"
        ),

        }

    g = sess.load_from(edges, vertices)
    lg = sess.learning(
    g,
    nodes=[("u", ["idfeature:string", "degree:int64","sqrtdegree:double"]),
    ("i", ["idfeature:string", "degree:int64","sqrtdegree:double"])],
    edges=[(("u", "u-i", 'i'), ["src_degree:int64", "dst_degree:int64", "src_sqrtdegree:double", "dst_sqrtdegree:double",],),
    (("i", "u-i_reverse", 'u'), [ "src_degree:int64", "dst_degree:int64", "src_sqrtdegree:double", "dst_sqrtdegree:double",],),
    (("u", "test-u-i", 'i'), ["src_degree:int64", "dst_degree:int64", "src_sqrtdegree:double", "dst_sqrtdegree:double",],),
    (("i", "test-u-i_reverse", 'u'), [ "src_degree:int64", "dst_degree:int64", "src_sqrtdegree:double", "dst_sqrtdegree:double",],),
    ],
    )
    try:
        train(config, lg)
    except:
        sess.close()
    sess.close()
```

**Log Output**
```bash
2021-07-06 19:27:48.906025: W tensorflow/core/framework/op_kernel.cc:1639] Invalid argument: TypeError: `generator` yielded an element that did not match the expected structure. The expected structure was (tf.int64, tf.float32, tf.string, tf.int64, tf.float32, tf.string, tf.int64, tf.float32, tf.string, tf.int64, tf.float32, tf.string, tf.int64, tf.float32, tf.string, tf.int64, tf.float32, tf.string, tf.int64, tf.int32, tf.float32), but the yielded element was (array([ 0, 24,  1, ..., 24,  8,  1]), array([[0.0000000e+00, 5.3000000e+01, 7.2801099e+00],
       [2.4000000e+01, 2.7850000e+03, 5.2773098e+01],
       [1.0000000e+00, 3.2800000e+02, 1.8110771e+01],
       ...,
       [2.4000000e+01, 2.7850000e+03, 5.2773098e+01],
       [8.0000000e+00, 4.4000000e+01, 6.6332498e+00],
       [1.0000000e+00, 3.2800000e+02, 1.8110771e+01]], dtype=float32), array([   29, 36281,    42, ..., 15383, 20348,   269]), array([[2.9000000e+01, 9.7000000e+01, 9.8488579e+00],
       [3.6281000e+04, 1.6000000e+01, 4.0000000e+00],
       [4.2000000e+01, 4.6000000e+01, 6.7823300e+00],
       ...,
       [1.5383000e+04, 3.5000000e+02, 1.8708286e+01],
       [2.0348000e+04, 8.6000000e+01, 9.2736187e+00],
       [2.6900000e+02, 2.9000000e+01, 5.3851647e+00]], dtype=float32), array([43818, 20346, 28859, ...,   752, 12175,  2713]), array([[4.3818000e+04, 9.0000000e+01, 9.4868326e+00],
       [2.0346000e+04, 2.5000000e+01, 5.0000000e+00],
       [2.8859000e+04, 1.3100000e+02, 1.1445523e+01],
       ...,
       [7.5200000e+02, 4.8000000e+01, 6.9282031e+00],
       [1.2175000e+04, 3.2400000e+02, 1.8000000e+01],
       [2.7130000e+03, 1.2500000e+02, 1.1180340e+01]], dtype=float32), array([   18,  2831, 14457, ...,  3552,  1171,    75]), array([[1.8000000e+01, 4.7000000e+01, 6.8556547e+00],
       [2.8310000e+03, 6.3000000e+01, 7.9372540e+00],
       [1.4457000e+04, 8.5000000e+01, 9.2195444e+00],
       ...,
       [3.5520000e+03, 4.8000000e+01, 6.9282031e+00],
       [1.1710000e+03, 1.6000000e+01, 4.0000000e+00],
       [7.5000000e+01, 1.8400000e+02, 1.3564660e+01]], dtype=float32), array([ 4180,  4486, 27810, ..., 22480, 18571, 12688]), array([[4.1800000e+03, 8.0000000e+01, 8.9442720e+00],
       [4.4860000e+03, 2.4000000e+01, 4.8989797e+00],
       [2.7810000e+04, 2.7000000e+01, 5.1961522e+00],
       ...,
       [2.2480000e+04, 2.0200000e+02, 1.4212670e+01],
       [1.8571000e+04, 3.3000000e+01, 5.7445626e+00],
       [1.2688000e+04, 1.6800000e+02, 1.2961481e+01]], dtype=float32), array([36112, 14691, 18548, ..., 13522, 16386, 32036]), array([[3.6112000e+04, 2.3000000e+01, 4.7958317e+00],
       [1.4691000e+04, 1.8000000e+01, 4.2426405e+00],
       [1.8548000e+04, 2.8000000e+01, 5.2915025e+00],
       ...,
       [1.3522000e+04, 2.0200000e+02, 1.4212670e+01],
       [1.6386000e+04, 1.9000000e+01, 4.3588991e+00],
       [3.2036000e+04, 1.6000000e+01, 4.0000000e+00]], dtype=float32), array([  18, 5410,  549, ..., 6175, 1573,  138]), array([1, 1, 0, ..., 1, 0, 1], dtype=int32), array([[  53.       ,   10.       ,    7.28011  ,    3.1622777],
       [2785.       ,   36.       ,   52.773098 ,    6.       ],
       [ 328.       ,   18.       ,   18.110771 ,    4.2426405],
       ...,
       [2785.       ,   46.       ,   52.773098 ,    6.78233  ],
       [  44.       ,   11.       ,    6.6332498,    3.3166249],
       [ 328.       ,   62.       ,   18.110771 ,    7.8740077]],
      dtype=float32)).
Traceback (most recent call last):

  File "/home/liushaorong/.conda/envs/py37/lib/python3.7/site-packages/tensorflow_core/python/data/ops/dataset_ops.py", line 600, in generator_py_func
    flattened_values = nest.flatten_up_to(output_types, values)

  File "/home/liushaorong/.conda/envs/py37/lib/python3.7/site-packages/tensorflow_core/python/data/util/nest.py", line 396, in flatten_up_to
    assert_shallow_structure(shallow_tree, input_tree)

  File "/home/liushaorong/.conda/envs/py37/lib/python3.7/site-packages/tensorflow_core/python/data/util/nest.py", line 311, in assert_shallow_structure
    % (len(input_tree), len(shallow_tree)))

ValueError: The two structures don't have the same sequence length. Input structure has length 15, while shallow structure has length 21.


During handling of the above exception, another exception occurred:


Traceback (most recent call last):

  File "/home/liushaorong/.conda/envs/py37/lib/python3.7/site-packages/tensorflow_core/python/ops/script_ops.py", line 235, in __call__
    ret = func(*args)

  File "/home/liushaorong/.conda/envs/py37/lib/python3.7/site-packages/tensorflow_core/python/data/ops/dataset_ops.py", line 605, in generator_py_func
    "element was %s." % (output_types, values))

TypeError: `generator` yielded an element that did not match the expected structure. The expected structure was (tf.int64, tf.float32, tf.string, tf.int64, tf.float32, tf.string, tf.int64, tf.float32, tf.string, tf.int64, tf.float32, tf.string, tf.int64, tf.float32, tf.string, tf.int64, tf.float32, tf.string, tf.int64, tf.int32, tf.float32), but the yielded element was (array([ 0, 24,  1, ..., 24,  8,  1]), array([[0.0000000e+00, 5.3000000e+01, 7.2801099e+00],
       [2.4000000e+01, 2.7850000e+03, 5.2773098e+01],
       [1.0000000e+00, 3.2800000e+02, 1.8110771e+01],
       ...,
       [2.4000000e+01, 2.7850000e+03, 5.2773098e+01],
       [8.0000000e+00, 4.4000000e+01, 6.6332498e+00],
       [1.0000000e+00, 3.2800000e+02, 1.8110771e+01]], dtype=float32), array([   29, 36281,    42, ..., 15383, 20348,   269]), array([[2.9000000e+01, 9.7000000e+01, 9.8488579e+00],
       [3.6281000e+04, 1.6000000e+01, 4.0000000e+00],
       [4.2000000e+01, 4.6000000e+01, 6.7823300e+00],
       ...,
       [1.5383000e+04, 3.5000000e+02, 1.8708286e+01],
       [2.0348000e+04, 8.6000000e+01, 9.2736187e+00],
       [2.6900000e+02, 2.9000000e+01, 5.3851647e+00]], dtype=float32), array([43818, 20346, 28859, ...,   752, 12175,  2713]), array([[4.3818000e+04, 9.0000000e+01, 9.4868326e+00],
       [2.0346000e+04, 2.5000000e+01, 5.0000000e+00],
       [2.8859000e+04, 1.3100000e+02, 1.1445523e+01],
       ...,
       [7.5200000e+02, 4.8000000e+01, 6.9282031e+00],
       [1.2175000e+04, 3.2400000e+02, 1.8000000e+01],
       [2.7130000e+03, 1.2500000e+02, 1.1180340e+01]], dtype=float32), array([   18,  2831, 14457, ...,  3552,  1171,    75]), array([[1.8000000e+01, 4.7000000e+01, 6.8556547e+00],
       [2.8310000e+03, 6.3000000e+01, 7.9372540e+00],
       [1.4457000e+04, 8.5000000e+01, 9.2195444e+00],
       ...,
       [3.5520000e+03, 4.8000000e+01, 6.9282031e+00],
       [1.1710000e+03, 1.6000000e+01, 4.0000000e+00],
       [7.5000000e+01, 1.8400000e+02, 1.3564660e+01]], dtype=float32), array([ 4180,  4486, 27810, ..., 22480, 18571, 12688]), array([[4.1800000e+03, 8.0000000e+01, 8.9442720e+00],
       [4.4860000e+03, 2.4000000e+01, 4.8989797e+00],
       [2.7810000e+04, 2.7000000e+01, 5.1961522e+00],
       ...,
       [2.2480000e+04, 2.0200000e+02, 1.4212670e+01],
       [1.8571000e+04, 3.3000000e+01, 5.7445626e+00],
       [1.2688000e+04, 1.6800000e+02, 1.2961481e+01]], dtype=float32), array([36112, 14691, 18548, ..., 13522, 16386, 32036]), array([[3.6112000e+04, 2.3000000e+01, 4.7958317e+00],
       [1.4691000e+04, 1.8000000e+01, 4.2426405e+00],
       [1.8548000e+04, 2.8000000e+01, 5.2915025e+00],
       ...,
       [1.3522000e+04, 2.0200000e+02, 1.4212670e+01],
       [1.6386000e+04, 1.9000000e+01, 4.3588991e+00],
       [3.2036000e+04, 1.6000000e+01, 4.0000000e+00]], dtype=float32), array([  18, 5410,  549, ..., 6175, 1573,  138]), array([1, 1, 0, ..., 1, 0, 1], dtype=int32), array([[  53.       ,   10.       ,    7.28011  ,    3.1622777],
       [2785.       ,   36.       ,   52.773098 ,    6.       ],
       [ 328.       ,   18.       ,   18.110771 ,    4.2426405],
       ...,
       [2785.       ,   46.       ,   52.773098 ,    6.78233  ],
       [  44.       ,   11.       ,    6.6332498,    3.3166249],
       [ 328.       ,   62.       ,   18.110771 ,    7.8740077]],
      dtype=float32)).


2021-07-06 19:27:54,759 [INFO][coordinator:603]: Coordinator close learning instance with object id 7195381417810821
E0706 19:27:58.413465523 3619820 backup_poller.cc:119]       run_poller: {"created":"@1625570878.413230554","description":"Timer list shutdown","file":"/tmp/grpc/src/core/lib/iomgr/timer_generic.cc","file_line":297}

```

**Comment by acezen** [comment]
Load feature with string type can use categorical_attrs_desc
```python
    vertices = {
        "u": (Loader("/home/jovyan/datasets/amazon-book/scope_user.csv", delimiter=",",), [("idfeature:string", "str"), "degree:int64","sqrtdegree:double"], ),
        "i": (Loader("/home/jovyan/datasets/amazon-book/scope_item.csv", delimiter=","), [("idfeature:string", "str"), "degree:int64","sqrtdegree:double"], ),
        }
    edges = {
        "u-i": (
            Loader("/home/jovyan/datasets/amazon-book/scope_train.csv", delimiter=","),
            [ "src_degree:int64", "dst_degree:int64", "src_sqrtdegree:double", "dst_sqrtdegree:double", "label"],
            ("src_id:int64", "u"),
            ("dst_id:int64", "i"),
            # "label:int64"
        ),
        "u-i_reverse": (
            Loader("/home/jovyan/datasets/amazon-book/scope_train.csv", delimiter=","),
            [ "src_degree:int64", "dst_degree:int64", "src_sqrtdegree:double", "dst_sqrtdegree:double", "label"],
            ("dst_id:int64", "i"),
            ("src_id:int64", "u"),
            # "label:int64"
        ),
        "test-u-i": (
            Loader("/home/jovyan/datasets/amazon-book/scope_train.csv", delimiter=","),
            [ "src_degree:int64", "dst_degree:int64", "src_sqrtdegree:double", "dst_sqrtdegree:double", "label"],
            ("src_id:int64", "u"),
            ("dst_id:int64", "i"),
            # "label:int64"
        ),
        "test-u-i_reverse": (
            Loader("/home/jovyan/datasets/amazon-book/scope_train.csv", delimiter=","),
            [ "src_degree:int64", "dst_degree:int64", ("src_sqrtdegree:double") "dst_sqrtdegree:double", "label"],
            ("dst_id:int64", "i"),
            ("src_id:int64", "u"),
            # "label:int64"
        ),

        }

    g = sess.load_from(edges, vertices)
```

**Comment by acezen** [comment]
pull request #686  fix the learning engine can't launch with lazy mode.

**Comment by acezen** [comment]
fixes a "get attribute" bug in GLE with pr #687
